File: collect_app/src/main/java/org/odk/collect/android/configure/qr/QRCodeScannerFragment.java
package org.odk.collect.android.configure.qr;

import android.content.Context;

import com.google.zxing.integration.android.IntentIntegrator;
import com.journeyapps.barcodescanner.BarcodeResult;

import org.odk.collect.android.R;
import org.odk.collect.android.activities.MainMenuActivity;
import org.odk.collect.analytics.Analytics;
import org.odk.collect.android.analytics.AnalyticsEvents;
import org.odk.collect.android.configure.SettingsImporter;
import org.odk.collect.android.fragments.BarCodeScannerFragment;
import org.odk.collect.android.injection.DaggerUtils;
import org.odk.collect.android.utilities.ToastUtils;
import org.odk.collect.shared.strings.Md5;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.util.Collection;
import java.util.Collections;
import java.util.zip.DataFormatException;

import javax.inject.Inject;

import static org.odk.collect.android.activities.ActivityUtils.startActivityAndCloseAllOthers;
import static org.odk.collect.android.utilities.CompressionUtils.decompress;

public class QRCodeScannerFragment extends BarCodeScannerFragment {

    @Inject
    SettingsImporter settingsImporter;

    @Inject
    Analytics analytics;

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        DaggerUtils.getComponent(context).inject(this);
    }

    @Override
    protected void handleScanningResult(BarcodeResult result) throws IOException, DataFormatException {
        boolean importSuccess = settingsImporter.fromJSON(decompress(result.getText()));
        String settingsHash = Md5.getMd5Hash(new ByteArrayInputStream(result.getText().getBytes()));

        if (importSuccess) {
            ToastUtils.showLongToast(getString(R.string.successfully_imported_settings));
            analytics.logEvent(AnalyticsEvents.SETTINGS_IMPORT_QR, "Success", settingsHash);
            startActivityAndCloseAllOthers(requireActivity(), MainMenuActivity.class);
        } else {
            ToastUtils.showLongToast(getString(R.string.invalid_qrcode));
            analytics.logEvent(AnalyticsEvents.SETTINGS_IMPORT_QR, "No valid settings", settingsHash);
        }
    }

    @Override
    protected Collection<String> getSupportedCodeFormats() {
        return Collections.singletonList(IntentIntegrator.QR_CODE);
    }
}

File: collect_app/src/main/java/org/odk/collect/android/configure/qr/QRCodeScannerFragment.kt
package org.odk.collect.android.configure.qr

import android.content.Context
import org.odk.collect.analytics.Analytics
import org.odk.collect.android.activities.ActivityUtils
import org.odk.collect.android.analytics.AnalyticsEvents
import org.odk.collect.android.fragments.BarCodeScannerFragment
import org.odk.collect.android.injection.DaggerUtils
import org.odk.collect.android.mainmenu.MainMenuActivity
import org.odk.collect.android.projects.ProjectsDataService
import org.odk.collect.android.storage.StoragePathProvider
import org.odk.collect.androidshared.ui.ToastUtils.showLongToast
import org.odk.collect.androidshared.ui.ToastUtils.showShortToast
import org.odk.collect.androidshared.utils.CompressionUtils
import org.odk.collect.projects.ProjectConfigurationResult
import org.odk.collect.settings.ODKAppSettingsImporter
import org.odk.collect.strings.R
import java.io.File
import javax.inject.Inject

class QRCodeScannerFragment : BarCodeScannerFragment() {

    @Inject
    lateinit var settingsImporter: ODKAppSettingsImporter

    @Inject
    lateinit var projectsDataService: ProjectsDataService

    @Inject
    lateinit var storagePathProvider: StoragePathProvider

    override fun onAttach(context: Context) {
        super.onAttach(context)
        DaggerUtils.getComponent(context).inject(this)
    }

    override fun handleScanningResult(result: String) {
        val oldProjectName = projectsDataService.requireCurrentProject().name

        try {
            val settingsImportingResult = settingsImporter.fromJSON(
                CompressionUtils.decompress(result),
                projectsDataService.requireCurrentProject()
            )

            when (settingsImportingResult) {
                ProjectConfigurationResult.SUCCESS -> {
                    Analytics.log(AnalyticsEvents.RECONFIGURE_PROJECT)

                    val newProjectName = projectsDataService.requireCurrentProject().name
                    if (newProjectName != oldProjectName) {
                        File(storagePathProvider.getProjectRootDirPath() + File.separator + oldProjectName).delete()
                        File(storagePathProvider.getProjectRootDirPath() + File.separator + newProjectName).createNewFile()
                    }

                    showLongToast(
                        getString(R.string.successfully_imported_settings)
                    )
                    ActivityUtils.startActivityAndCloseAllOthers(
                        requireActivity(),
                        MainMenuActivity::class.java
                    )
                }

                ProjectConfigurationResult.INVALID_SETTINGS -> {
                    showLongToast(
                        getString(
                            R.string.invalid_qrcode
                        )
                    )
                    restartScanning()
                }

                ProjectConfigurationResult.GD_PROJECT -> {
                    showLongToast(
                        getString(R.string.settings_with_gd_protocol)
                    )
                    restartScanning()
                }
            }
        } catch (e: Exception) {
            showShortToast(getString(R.string.invalid_qrcode))
            restartScanning()
        }
    }

    override fun isQrOnly(): Boolean {
        return true
    }
}
